test/e2e/framework/kubectl/kubectl_builder.go

```diff
@@ -0,0 +1,345 @@
+/*
+Copyright 2020 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package kubectl
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"io"
+	"net"
+	"net/url"
+	"os/exec"
+	"strings"
+	"syscall"
+	"time"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/util/wait"
+	clientset "k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/tools/clientcmd"
+	"k8s.io/kubernetes/test/e2e/framework"
+	uexec "k8s.io/utils/exec"
+)
+
+// KubectlBuilder is used to build, customize and execute a kubectl Command.
+// Add more functions to customize the builder as needed.
+type KubectlBuilder struct {
+	cmd     *exec.Cmd
+	timeout <-chan time.Time
+}
+
+// NewKubectlCommand returns a KubectlBuilder for running kubectl.
+func NewKubectlCommand(namespace string, args ...string) *KubectlBuilder {
+	b := new(KubectlBuilder)
+	tk := NewTestKubeconfig(framework.TestContext.CertDir, framework.TestContext.Host, framework.TestContext.KubeConfig, framework.TestContext.KubeContext, framework.TestContext.KubectlPath, namespace)
```

Comment:
This is a bit of an antipattern that we have because it relies on the magic of the e2e framework which makes things hard to debug/track sometimes.
In terms of code organization, it would also be best to not import the test framework within one of its subpackages (specially if it is for passing this sort of data - later on its used for logging so :shrug: ).

For this, I think, the best thing to do would be to refactor it in such a way that `e2ekubectl.NewTestKubeconfig` was built in the code calling `e2ekubectl` and then the `e2ekubectl.TestKubeconfig` (https://github.com/kubernetes/kubernetes/blob/master/test/e2e/framework/kubectl/kubectl_utils.go) instance was passed around to this function.
Ultimately changing the signature here to something like

```go
// NewKubectlCommand returns a KubectlBuilder for running kubectl.
func NewKubectlCommand(tk *TestKubeconfig) *KubectlBuilder {
  ...
```

Reply:
Makes sense to me.  Will do.